### PR TITLE
image viewer zoom-level/center

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request moves the logic for translating between zoom limits (in x and y) and zoom level/center from the astrowidgets API into the viewer state and will eventually also include a UI (building on #2604) to allow accessing and modifying these from the plot options plugin.

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by pllim — COMMENTED** [review]
You might wanna wait. I redid some of the math to account for Brett's rotation work on his branch.

**Comment by kecnry** [comment]
> You might wanna wait.

I think that should be pretty easy to migrate here, no?

**Comment by pllim** [comment]
Yes. But you will have conflicts.

File: jdaviz/core/astrowidgets_api.py [hunk 1]
```diff
@@ -99,14 +99,7 @@ def center_on(self, point):
...
             pix = point
 
-        width = self.state.x_max - self.state.x_min
-        height = self.state.y_max - self.state.y_min
-
-        with delay_callback(self.state, 'x_min', 'x_max', 'y_min', 'y_max'):
-            self.state.x_min = pix[0] - (width * 0.5)
-            self.state.y_min = pix[1] - (height * 0.5)
-            self.state.x_max = self.state.x_min + width
-            self.state.y_max = self.state.y_min + height
```
Comment (kecnry) [inline jdaviz/core/astrowidgets_api.py:109]: I ended up rewriting this logic since I forgot it already existed... I'll have to think which way is more reliable or if there is any actual difference 🤔

File: jdaviz/core/astrowidgets_api.py [hunk 2]
```diff
@@ -196,19 +184,8 @@ def zoom_level(self, val):
...
             return
-        else:
-            cur_xcen = (self.state.x_min + self.state.x_max) * 0.5
-            new_dx = self.shape[1] * 0.5 / val
-            new_x_min = cur_xcen - new_dx
-            new_x_max = cur_xcen + new_dx
-
-        with delay_callback(self.state, 'x_min', 'x_max'):
-            self.state.x_min = new_x_min - 0.5
-            self.state.x_max = new_x_max - 0.5
```
Comment (kecnry) [inline jdaviz/core/astrowidgets_api.py:207]: @pllim should this be `+ 0.5` instead (assuming the 0.5 is to account for the margin added around the data)?  Otherwise, where is this coming from?

Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:207]: I don't quite remember either and https://github.com/spacetelescope/jdaviz/pull/744 has no record of it. I think it has something to do with user input being pixel center but viewer wants pixel corner. Not 100% sure.

Comment (pllim) [reply jdaviz/core/astrowidgets_api.py:207]: I was sure I worked out the math properly though using API + mouse cursor info. Does it not work anymore?

Comment (kecnry) [reply jdaviz/core/astrowidgets_api.py:207]: I was seeing wandering when changing the zoom-level and switching to `+` seemed to fix it, but maybe introduced that somewhere else.  I'll go back to main and see if I can trigger the same there or not.  Either way, doesn't _seem_ to be the culprit for the failing test.

**Comment by pllim** [comment]
Just focusing on the first failure I see, looks like centering failed. You have `array([-0.5,  9.5, -0.5,  9.5])` which means it is still centering on the image center, not the `(0, 1)` I told it to go to.

```
    def test_center_offset_pixel(self):
        self.viewer.center_on((0, 1))
>       assert_allclose((self.viewer.state.x_min, self.viewer.state.x_max,
                         self.viewer.state.y_min, self.viewer.state.y_max),
                        (-5, 5, -4, 6))
```

**Comment by pllim** [comment]
Try run the test case interactively on a notebook. Does it work there?

**Comment by kecnry** [comment]
Some of the failures here I have a hacky workaround locally, I just need to figure out the right way to do it, but there are a few others that still remain after that as well.

File: jdaviz/core/freezable_state.py [hunk 3]
```diff
@@ -97,15 +179,14 @@ def reset_limits(self, *event):
                 if not layer.visible or layer.layer.data.ndim == 1:
                     continue
                 pixel_ids = layer.layer.pixel_component_ids
+                pixel_id_x = [comp for comp in pixel_ids if comp.label.endswith('[x]')][0]
+                pixel_id_y = [comp for comp in pixel_ids if comp.label.endswith('[y]')][0]
 
-                x_max = max(x_max, layer.layer.data[pixel_ids[1]].max() + 0.5)
-                y_max = max(y_max, layer.layer.data[pixel_ids[0]].max() + 0.5)
+                x_max = max(x_max, layer.layer.data[pixel_id_x].max() + 0.5)
+                y_max = max(y_max, layer.layer.data[pixel_id_y].max() + 0.5)
```
Comment (kecnry) [inline jdaviz/core/freezable_state.py:186]: this technically isn't necessary, but will help future-proof if we ever do enable this custom `reset_limits` for any cube viewer.

**Comment by pllim** [comment]
This is the only error I am seeing:

`AttributeError: type object 'BqplotImageViewerState' has no attribute 'zoom_center_x'`

File: jdaviz/core/freezable_state.py (hunk 3 above)
Comment (pllim) [reply jdaviz/core/freezable_state.py:187]: But now it depends on the label name. Is the label always guaranteed to have x/y in the name?

File: jdaviz/configs/cubeviz/plugins/viewers.py [hunk 4]
```diff
@@ -31,9 +32,12 @@ class CubevizImageView(JdavizViewerMixin, BqplotImageView):
                 ]
 
     default_class = None
+    _state_cls = FreezableBqplotImageViewerState
```
[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/viewers.py:35]: No, I meant `FreezableBqplotImageViewerState.reset_limits()` which overrides glue's `BqplotImageViewerState.reset_limits()`

File: jdaviz/core/freezable_state.py (hunk 3 above)
Comment (kecnry) [reply jdaviz/core/freezable_state.py:187]: that seems to be more general/reliable than assuming the first and second index 🤷‍♂️   I can revert for now though and deal with this if/when we try to make the image viewers in configs more consistent.

**Comment by kecnry** [comment]
That is the mosviz error I am still working on addressing.  If you remove the if-statement within `FreezableBqplotImageViewerState.reset_limits()` then you'll see the cubeviz tests fail because of the different behavior between that and the upstream `reset_limits()`.  I fixed it for now by just still sending cubeviz to the upstream implementation since it feels out of scope here to change the zoom behavior itself that would require a change to the tests.

File: jdaviz/configs/cubeviz/plugins/viewers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/viewers.py:35]: Looks like you implemented this 2 years ago in https://github.com/spacetelescope/jdaviz/pull/1897

[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/viewers.py:35]: I know, and that so far has not been enabled in cubeviz.

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/viewers.py:35]: And then that particular line was changed from a different logic by me later in https://github.com/spacetelescope/jdaviz/pull/1928

**Review by javerbukh — APPROVED** [review]
Zooming and centering work as expected, even when viewers are linked. Plot options also handles different combinations of multiselect well, even when an image and spectrum viewer are selected. Nice work!

**Review by cshanahan1 — COMMENTED** [review]
Zooming in looks ok at first, but it starts to jump around and drift eventually: 

https://github.com/spacetelescope/jdaviz/assets/20444899/56896474-d4e2-41d0-8472-6c8ac67dabb3

**Comment by cshanahan1** [comment]
I first select uncert viewer, and change the zoom level. That works. Then I change to flux viewer. If I press 'reset' it will reset the flux viewer to some initial zoom level (even though it hasn't been modified...?), but the zoom level changes in the uncert viewer instead of the flux viewer for some reason.

Uploading Screen Recording 2024-01-11 at 4.57.43 PM.mov…

**Comment by cshanahan1** [comment]
A general comment I have - I think the plot options plugin is a bit cluttered and could benefit from some reorganization. The addition of this in a dropdown within the plugin makes sense, so I'm not sure what I would do to improve that within this PR, but with the other options it all feels like its not organized in the most logical way.

**Comment by kecnry** [comment]
> Zooming in looks ok at first, but it starts to jump around and drift eventually:

I _think_ this is due to the way that the zoom-level is processed in order to maintain the aspect-ratio and exists in the astrowidgets implementation as well, so are you okay with marking this as a follow-up effort since fixing it would change behavior there?

> I first select uncert viewer, and change the zoom level. That works. Then I change to flux viewer. If I press 'reset' it will reset the flux viewer to some initial zoom level (even though it hasn't been modified...?), but the zoom level changes in the uncert viewer instead of the flux viewer for some reason.

I cannot reproduce and your video didn't post.  Can you post that again?  Is it possible you had a linked zoom tool enabled when clicking the reset button?

**Comment by cshanahan1** [comment]
just updated my comment with the video

**Comment by kecnry** [comment]
Moved back to draft while I update this to work on top of image rotation now that #2179 is merged.  This likely will result in exposing zoom-radius instead of zoom-level to the user in plot options, while keeping the astrowidgets PR intact.  The definition for zoom-level in when WCS-linked, in a way that is astrowidgets-consistent, can then be revisited as a follow-up effort.

**Comment by kecnry** [comment]
@javerbukh - since your approval, this has been updated to expose zoom-radius instead of zoom-level (with different translation logic), but the UI remains essentially unchanged.  Feel free to dismiss your review though if you see anything in the latest two commits that you think warrants a fresh look.

**Review by bmorris3 — COMMENTED** [review]
Something is amiss in Imviz w/o WCS linking. Changing zoom radius in the plugin causes jumping in the viewer, and does not change the zoom radius.


https://github.com/spacetelescope/jdaviz/assets/3497584/82436c71-562d-44db-8109-0924b75b7e98

**Comment by bmorris3** [comment]
When linked by WCS, zoom radius step size of "1" (degree?) is too big to be useful.

In some other Plot Options context (stretch histogram maybe?), we adjust the up/down arrow step size depending on the content of the viewer. I think we need that here. 

Also, we need some kind of unit on the zoom center (x, y) and radius. When WCS-linked, it's degrees for x and y, and maybe also radius. I'd suggest that zoom radius should be displayed in units of arcsec/arcmin.

https://github.com/spacetelescope/jdaviz/assets/3497584/9a9c3195-8bd4-4c8a-8d4f-0c568efc8264

File: jdaviz/core/freezable_state.py [hunk 5]
```diff
@@ -54,11 +57,92 @@ def _reset_x_limits(self, *event):
...
+        if self.zoom_radius <= 0.0:
+            raise ValueError("zoom_radius must be positive")
+
+        # When WCS-linked (displayed on the sky): zoom_center_x/y and zoom_radius are in sky units,
+        # x/y_min/max are in pixels of the WCS-only layer
+        if self.linked_by_wcs:
+            image, i_ref = get_reference_image_data(self._viewer.jdaviz_app, self._viewer.reference)
+            ref_wcs = image.coords
+            center_x, center_y = ref_wcs.world_to_pixel_values(self.zoom_center_x, self.zoom_center_y)  # noqa
+            center_xr, center_yr = ref_wcs.world_to_pixel_values(self.zoom_center_x+self.zoom_radius, self.zoom_center_y)  # noqa
```
Comment (pllim) [inline jdaviz/core/freezable_state.py:96]: I think it is safer to stick to high-level API from https://github.com/astropy/astropy-APEs/blob/main/APE14.rst#high-level-api instead of using the low-level API. Also, I think both `world_to_pixel` and `world_to_pixel_values` can take multiple inputs at once, so it is more performant to vectorize both `center` and `center + radius` inputs in just one call.

File: jdaviz/core/freezable_state.py [hunk 6]
```diff
@@ -54,11 +57,92 @@ def _reset_x_limits(self, *event):
...
+        if None in (self.x_min, self.x_max, self.y_min, self.y_max):
+            return
+
+        # When WCS-linked (displayed on the sky): zoom_center_x/y and zoom_radius are in sky units,
+        # x/y_min/max are in pixels of the WCS-only layer
+        if self.linked_by_wcs:
+            image, i_ref = get_reference_image_data(self._viewer.jdaviz_app, self._viewer.reference)
+            ref_wcs = image.coords
+            x_min, y_min = ref_wcs.pixel_to_world_values(self.x_min, self.y_min)
+            x_max, y_max = ref_wcs.pixel_to_world_values(self.x_max, self.y_max)
```
Comment (pllim) [inline jdaviz/core/freezable_state.py:130]: Same comment as https://github.com/spacetelescope/jdaviz/pull/2649/files#r1476525669

**Review by pllim — COMMENTED** [review]
Since changes to Astrowidgets API so far is minimal, I do not have concern on the design. When you have made more progress, the user docs definitely would need to be updated. Also does not hurt to update the example notebook(s) to demonstrate this new way of zooming.

Thanks!

**Comment by kecnry** [comment]
> Something is amiss in Imviz w/o WCS linking. Changing zoom radius in the plugin causes jumping in the viewer, and does not change the zoom radius.

I think the zoom radius was working, it just isn't obvious with the stepsize.  Dynamic step-size is now implemented, but you'll need https://github.com/glue-viz/glue-jupyter/pull/417 in order to see it in action (without that upstream PR, the stepsize will fallback on 1).

Jumping is as fixed as I can get it 😬 

> When linked by WCS, zoom radius step size of "1" (degree?) is too big to be useful, and should be restricted to positive values.
> In some other Plot Options context (stretch histogram maybe?), we adjust the up/down arrow step size depending on the content of the viewer. I think we need that here.

Re step size: see response above.

Re negative radius: limiting the value through glue is difficult, so I now just gracefully handle a negative radius 🤷‍♂️ 

> Also, we need some kind of unit on the zoom center (x, y) and radius. When WCS-linked, it's degrees for x and y, and maybe also radius. I'd suggest that zoom radius should be displayed in units of arcsec/arcmin.

For now I just expose these as degrees to avoid unit-conversion.  I can see the argument for making zoom-radius a "finer" unit, but suspect we'll want to just adhere to display units if/when we implement them for the spatial direction on images.

Still known todo items:
* fix broken tests
* add new test coverage
* see what docs need updating

File: jdaviz/core/freezable_state.py (hunk 5 above)
Comment (kecnry) [reply jdaviz/core/freezable_state.py]: Is it worth it to pass units around just to get rid of them though?  As long as the low-level API is considered stable, is there harm in using it?

As for combining - how would you suggest doing that so that this is still readable?

Comment (pllim) [reply jdaviz/core/freezable_state.py]: > Is it worth it to pass units around just to get rid of them though?

Ah, units. Okay, low-level is fine then.

> readable?

I think it would be something like this. Basically you do the look-up once and then unpack later. Look-up can be expensive but unpacking shouldn't be.

```python
coo = ref_wcs.world_to_pixel_values([a bunch of x], [a bunch of y])
center_x  = coo[1]
center_y = ...
```

Comment (kecnry) [reply jdaviz/core/freezable_state.py]: That's what I feared... its just so hard to parse when reading, but if you think its worth it from performance, I can do it

Comment (pllim) [reply jdaviz/core/freezable_state.py]: For 2 transformations, you could argue that performance does not matter but I cannot tell for sure how much this will be compounded. You can try to benchmark this call with the distorted case from `ImvizDitheredExample` and see.

Comment (kecnry) [reply jdaviz/core/freezable_state.py]: ok, I think I found a way to keep it somewhat readable, let me know what you think

File: jdaviz/core/freezable_state.py [hunk 7]
```diff
@@ -54,14 +57,90 @@ def _reset_x_limits(self, *event):
...
+
+        # When WCS-linked (displayed on the sky): zoom_center_x/y and zoom_radius are in sky units,
+        # x/y_min/max are in pixels of the WCS-only layer
+        if self.linked_by_wcs:
+            image, i_ref = get_reference_image_data(self._viewer.jdaviz_app, self._viewer.reference)
+            ref_wcs = image.coords
+            cr = ref_wcs.world_to_pixel_values((self.zoom_center_x, self.zoom_center_x+abs(self.zoom_radius)),  # noqa
+                                               (self.zoom_center_y, self.zoom_center_y))
+            center_x, center_xr = cr[0]
+            center_y, center_yr = cr[1]
```
Comment (pllim) [inline jdaviz/core/freezable_state.py:96]:
```suggestion
            center_y = cr[1][0]
```

File: jdaviz/core/freezable_state.py (hunk 5 above)
Comment (pllim) [reply jdaviz/core/freezable_state.py]: LGTM thanks!

File: jdaviz/core/freezable_state.py (hunk 7 above)
Comment (pllim) [reply jdaviz/core/freezable_state.py:96]: Might want to add a comment like this because I had to remind myself what the inputs and outputs actually mean because the astropy doc and APE 14 aren't very clear at a glance:

```python
array([x1, x2, ...]), array([y1, y2, ...]) = ref_wcs.world_to_pixel_values([ra1_deg, ra2_deg, ...], [dec1_deg, dec2_deg, ...])
```

Comment (pllim) [reply jdaviz/core/freezable_state.py:96]: Underscore is fine, because technically I don't think `center_yr` assignment is correct.

**Review by javerbukh — APPROVED** [review]
Re-checked the code and tested the functionality and all still looks good. Others have mentioned the jumping but I can understand that that may not be fixable at the moment. As long as we have a ticket to track that this should be good to merge, thanks!

**Review by cshanahan1 — APPROVED** [review]
tested it again, looks good to me
